### ISSUE TITLE
Parsers: Use the Error constructor instead of "try" and "throw"?

### ISSUE THREAD

**Issue by harendra-kumar** [composewell/streamly#586]
Consider this code in `splitMany`:

```
    -- XXX The "try" may impact performance if this parser is used as a scan
    extract (Tuple3' s _ fs) = do
        r <- try $ extract1 s
        case r of
            Left (_ :: ParseError) -> fextract fs
            Right b -> fstep fs b >>= fextract
```

We can possibly use the `Error` constructor for all error propagations. The above code could be very inefficient because of the `catch#` blocking fusion if we use the parser as a scan.

It is possible to use the `Error` constructor, in fact I used it earlier during the implementation of parsers but the code became more complicated. This issue is to try it again and see where can we go with this. Also we can remove the MonadCatch/MonadThrow constraints on many combinators with this?

**Comment by harendra-kumar** [comment]
See also #585 .

**Comment by pranaysashank** [comment]
https://github.com/composewell/streamly/issues/604

**Comment by harendra-kumar** [comment]
From #671 we could use the following type for `extract`:

```

data Extract s b =
      EPartial Int b -- partial result and unused input count
    | EDone Int b   -- final result with unused input count
    | EError String
```

**Comment by harendra-kumar** [comment]
fixed by #1782
